crates/node-maintainer/src/graph.rs

```diff
@@ -185,6 +197,249 @@ impl Graph {
+                if parent_idx != idx {
+                    current = parent_idx
+                }
+            }
+
+            targets.insert(DemotionTarget {
+                target_idx: current,
+                dependent_idx: d.source(),
+                edge_idx: d.id(),
+            });
+        }
+
+        return targets.drain().collect();
+    }
+
+    /// True if every Node's sub-dependency either:
+    /// - Has a parent whose is an ancestor of Node's parent
+    /// - Within Node's subtree (including the node itself)
+    ///
+    /// If either condition are true for every sub-dependency - node could be
+    /// made sibling of its parent (promoted).
+    pub(crate) fn can_be_promoted(&self, node: NodeIndex) -> bool {
+        let node_parent = self.inner[node].parent.expect("parent of non-root node");
+        let mut q = VecDeque::new();
+        q.push_back(node);
+
+        while let Some(dep) = q.pop_front() {
+            q.extend(self.inner.neighbors_directed(dep, Direction::Outgoing));
+
+            if dep == node {
+                continue;
+            }
+
+            if let Some(dep_parent) = self.inner[dep].parent {
+                // Dependency is a sibling of node, can't promote
+                if dep_parent == node_parent {
+                    return false;
+                }
+
+                if !self.is_ancestor(dep_parent, node_parent) && !self.is_ancestor(node, dep) {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    /// Merge `deduped` nodes into a single node and put it as child of `target`
+    pub(crate) fn merge_and_promote(&mut self, target: NodeIndex, deduped: &[NodeIndex]) {
+        // Take the first node from the `deduped` list.
+        let mut deduped = deduped.iter();
+        let promoted_idx = deduped
+            .next()
+            .expect("at least one item in deduped list")
+            .to_owned();
+
+        // Disconnect the node from its current parent.
+        let name = UniCase::new(self.inner[promoted_idx].package.name().to_string());
+        if let Some(parent) = self.inner[promoted_idx].parent {
+            let removed = self.inner[parent].children.remove(&name);
+            assert_eq!(removed, Some(promoted_idx));
+        }
+
+        // Promote the node by placing it as the child of the `target`.
+        self.inner[target].children.insert(name, promoted_idx);
+        self.inner[promoted_idx].parent = Some(target);
+
+        // Remove the rest of `deduped` from the graph
+        let mut dependents = Vec::new();
+        for &idx in deduped {
+            // Move all dependents to the saved node.
+            dependents.extend(
+                self.inner
+                    .edges_directed(idx, Direction::Incoming)
+                    .map(|r| (r.source(), r.id())),
+            );
+            for (source_idx, edge_idx) in dependents.drain(..) {
```

Comment:
You'd see this pattern below as well. Basically, my idea was that we could allocate and grow the vector once and then reuse it. Could be a premature optimization so I'll leave it to you whether you want to remove it before merging the PR.

Reply:
Why put stuff in a dependents array instead of just using the interior directly? Am I just reading things wrong?

Reply:
Because we are mutating the `self.inner` and `edges_dirrected` holds an immutable reference.